### PR TITLE
swarm/storage: Fix garbage collection related index corruption

### LINKED ISSUE #931: Connected swarm node deletes locally uploaded content
How to reproduce:

- start a clean node
- upload a directory with multiple files and subdirectories with multiple files (a web site) 
- wait for 1 - 2 days running the node online

After running for a couple of days chunks are lost. and even reuploading the content again does not help.

### PR DESCRIPTION
**Pull request by nolash**
This PR fixes the invalid byte slice passed to the access count incrementer, causing index pointers to persist even though data rows and gc rows persist.

It also introduces a Cleanup method for correcting the damage done to already existing datastores due to the above.

closes #931 

Tasks:

- [x] Fix corrupt data parameter to accesscount updater
- [x] Cleanup method for correcting all entry nodes
- [x] Add test for entry cleanups
- [x] Count remaining chunks and update entrycount and proximity bin pointers
- [x] Add test for summary entries
- [x] Add cleanup to migration
- [x] ~Merge `CleanIndex` with `Cleanup`~ moved to https://github.com/ethersphere/go-ethereum/issues/985

### CODE REVIEW

File: swarm/storage/ldbstore.go [hunk 1]
```diff
@@ -554,49 +562,133 @@ func (s *LDBStore) Cleanup(f func(*chunk) bool) {
...
+			batch.Put(poKey[:], idxCount[:])
+		}
+	}
+
+	s.db.Write(&batch)
+
+	return nil
 }
 
+// this seems to be unused
```
Comment (zelig) [inline swarm/storage/ldbstore.go]: lets remove then

File: swarm/storage/ldbstore.go [hunk 2]
```diff
@@ -719,6 +813,7 @@ func (s *LDBStore) doPut(chunk Chunk, index *dpaDBIndex, po uint8) {
 	dkey := getDataKey(s.dataIdx, po)
 	s.batch.Put(dkey, data)
 	index.Idx = s.dataIdx
+	// TODO: why is this indexcount and not the count of items in po
```
Comment (zelig) [inline swarm/storage/ldbstore.go]: because this is a cursor where left off with PULL syncing

**Review by zelig — COMMENTED** [review]
@ @

File: swarm/storage/localstore.go [hunk 3]
```diff
@@ -205,8 +205,12 @@ func (ls *LocalStore) Migrate() error {
 	if schema != CurrentDbSchema {
 		// run migrations
 
-		if schema == "" {
+		if schema != DbSchemaHalloween {
 			log.Debug("running migrations for", "schema", schema, "runtime-schema", CurrentDbSchema)
+			if schema == DbSchemaPurity {
+				// fix garbage screwed garbage collection and indices
```
Comment (zelig) [inline swarm/storage/localstore.go]: comment!

File: swarm/storage/localstore.go [hunk 4]
```diff
@@ -205,8 +205,12 @@ func (ls *LocalStore) Migrate() error {
 	if schema != CurrentDbSchema {
 		// run migrations
 
-		if schema == "" {
```
Comment (nonsense) [inline swarm/storage/localstore.go:208]: I find this confusing.

We should have an if-block for every type of dbschema that is possible on users machines.

It is possible that a user has not updated their Swarm process and are still on schema == "", it is possible that they are also on schema purity and it is possible that they have updated to halloween - this makes it 3 blocks.

Having nested if checks for this is not ideal.

File: swarm/storage/schema.go [hunk 5]
```diff
@@ -2,5 +2,7 @@ package storage
 
 // "purity" is the first formal schema of LevelDB we release together with Swarm 0.3.5
 const DbSchemaPurity = "purity"
+const DbSchemaHalloween = "halloween"
```
Comment (nonsense) [inline swarm/storage/schema.go]: Please add a comment about this schema, and why/when we are introducing it.

File: swarm/storage/schema.go [hunk 6]
```diff
@@ -2,5 +2,7 @@ package storage
...
 
-const CurrentDbSchema = DbSchemaPurity
+//const CurrentDbSchema = DbSchemaPurity
```
Comment (nonsense) [inline swarm/storage/schema.go]: Remove commented code please.

File: swarm/storage/ldbstore_test.go [hunk 7]
```diff
@@ -584,6 +585,140 @@ func TestLDBStoreCollectGarbageAccessUnlikeIndex(t *testing.T) {
...
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// remove the data of the first chunk
+	po := ldb.po(chunks[0].Address()[:])
+	dataKey := make([]byte, 10)
+	dataKey[0] = keyData
+	dataKey[1] = byte(po)
+	_, err = ldb.db.Get(dataKey)
```
Comment (zelig) [inline swarm/storage/ldbstore_test.go:606]: this is confusing maybe worth a comment that the first chunk has storageIdx 0 on `dataKey[2:10]`

File: swarm/storage/ldbstore.go [hunk 8]
```diff
@@ -692,6 +737,7 @@ func (s *LDBStore) Put(ctx context.Context, chunk Chunk) error {
 	index.Access = s.accessCnt
 	s.accessCnt++
 	idata = encodeIndex(&index)
+	log.Warn("inserting index value", "v", idata, "index", index)
```
Comment (zelig) [inline swarm/storage/ldbstore.go]: remove logline

File: swarm/storage/ldbstore.go [hunk 9]
```diff
@@ -664,6 +708,7 @@ func (s *LDBStore) CurrentStorageIndex() uint64 {
 // Put adds a chunk to the database, adding indices and incrementing global counters.
 // If it already exists, it merely increments the access count of the existing entry.
 // Is thread safe
+
```
Comment (zelig) [inline swarm/storage/ldbstore.go]: remove newline

File: swarm/storage/ldbstore.go [hunk 10]
```diff
@@ -554,47 +558,87 @@ func (s *LDBStore) Cleanup(f func(*chunk) bool) {
 	log.Warn(fmt.Sprintf("Found %v errors out of %v entries. Removed %v chunks.", errorsFound, total, removed))
 }
 
-func (s *LDBStore) ReIndex() {
-	//Iterates over the database and checks that there are no faulty chunks
+func (s *LDBStore) CleanIndex() error {
```
Comment (zelig) [inline swarm/storage/ldbstore.go]: comment function

**Review by zelig — CHANGES_REQUESTED** [review]

**Review by zelig — APPROVED** [review]

**Comment by frncmx** [comment]
PR has been opened to upstream.
https://github.com/ethereum/go-ethereum/pull/18080

File: swarm/storage/localstore.go [hunk 11]
```diff
@@ -195,31 +195,42 @@ func (ls *LocalStore) Close() {
...
+	}
 
-			err := ls.DbStore.PutSchema(DbSchemaPurity)
-			if err != nil {
-				log.Error(err.Error())
-				return err
-			}
-		}
+	if actualDbSchema == DbSchemaPurity {
+		ls.DbStore.CleanGCIndex()
```
Comment (nolash) [inline swarm/storage/localstore.go]: Shouldn't this be `ls.migrateFromPurityToHalloween()`?

**Pull request closed without merging**